### ISSUE TITLE
Candidate extractors are outputting duplicates (Cross-platform Inconsistencies?)

### ISSUE THREAD

**Issue by lukehsiao** [lukehsiao/HardwareDeepDive#27]
When I view the entries in `part_collector_current_candidates`, I'm seeing a lot of duplicates, even though I'm using a set in `ext_part_temerature_candidates.py` to address this problem.

**Comment by lukehsiao** [comment]
One possibility is that there ar multiple ext_part_temperature_candidates extractors runnig in parallel, which circumvents using the set...

**Comment by xiao-cheng** [comment]
Running scripts on different files in parallel shouldn't be a problem, as the docid is included in the candidate string. Also make sure the candidates are all strings with the same encoding. if things still won't work, check hash(object) for the objects. I've had problem with floats before where they have equal value but different precision, resulting in different hash

**Comment by lukehsiao** [comment]
I'm trying to hash strings of the form of the output. For example:

```python
    candidates.add('\t'.join([str(i).strip() for i in[
                            x[PART_DOC_ID],
                            x[PART_NUM],
                            x[CC_MAX],
                            x[CC_UNIT],
                            rel_id,
                            is_true
                            ]]))
for candidate in candidates:
        print candidate
```

After `deepdive run` I get this:
```
deepdive sql 'select doc_id, part_num, collector_current_max, collector_current_unit, is_true from part_collector_current_candidates where is_true=true order by doc_id'

            doc_id            | part_num  | collector_current_max | collector_current_unit | is_true
------------------------------+-----------+-----------------------+------------------------+---------
 112823.pdf                   | BC546B    |                   100 | mA                     | t
 112823.pdf                   | BC547A    |                   100 | mA                     | t
 112823.pdf                   | BC546B    |                   100 | mA                     | t
 112823.pdf                   | BC547C    |                   100 | mA                     | t
 112823.pdf                   | BC548C    |                   100 | mA                     | t
 112823.pdf                   | BC548B    |                   100 | mA                     | t
 112823.pdf                   | BC548B    |                   100 | mA                     | t
 112823.pdf                   | BC546B    |                   100 | mA                     | t
 112823.pdf                   | BC547A    |                   100 | mA                     | t
```

Is there some naive python mistake I'm making?

**Comment by xiao-cheng** [comment]
the rel_id made these candidates different.
They are just the same part number occurring in multiple locations in the same pdf, I think.

**Comment by lukehsiao** [comment]
The rel_id is the same between them as well, though:
```
            doc_id            | part_num  | collector_current_max | collector_current_unit |                   relation_id                    | is_true
------------------------------+-----------+-----------------------+------------------------+--------------------------------------------------+---------
 112823.pdf                   | BC546B    |                   100 | mA                     | 112823.pdf~^~BC546B~^~100~^~mA                   | t
 112823.pdf                   | BC547A    |                   100 | mA                     | 112823.pdf~^~BC547A~^~100~^~mA                   | t
 112823.pdf                   | BC546B    |                   100 | mA                     | 112823.pdf~^~BC546B~^~100~^~mA                   | t
 112823.pdf                   | BC547C    |                   100 | mA                     | 112823.pdf~^~BC547C~^~100~^~mA                   | t
 112823.pdf                   | BC548C    |                   100 | mA                     | 112823.pdf~^~BC548C~^~100~^~mA                   | t
 112823.pdf                   | BC548B    |                   100 | mA                     | 112823.pdf~^~BC548B~^~100~^~mA                   | t
 112823.pdf                   | BC548B    |                   100 | mA                     | 112823.pdf~^~BC548B~^~100~^~mA                   | t
 112823.pdf                   | BC546B    |                   100 | mA                     | 112823.pdf~^~BC546B~^~100~^~mA                   | t
 112823.pdf                   | BC547A    |                   100 | mA                     | 112823.pdf~^~BC547A~^~100~^~mA                   | t
```

**Comment by xiao-cheng** [comment]
This is my run, which seemed fine:

```
xiao@DNab40468a ~/hdeepdive $ deepdive sql
psql (9.4.4)
Type "help" for help.

hwdeepdive=# select * from part_collector_current_candidates;                                                                                                                                                      doc_id   |  part_num  | collector_current_max | collector_current_unit |            relation_id             | is_true
------------+------------+-----------------------+------------------------+------------------------------------+---------
 112823.pdf | BC547A     |                   100 | mA                     | 112823.pdf~^~BC547A~^~100~^~mA     | t
 112823.pdf | BC547ARL   |                   100 | mA                     | 112823.pdf~^~BC547ARL~^~100~^~mA   |
 112823.pdf | BC546BZL1G |                   100 | mA                     | 112823.pdf~^~BC546BZL1G~^~100~^~mA |
 112823.pdf | BC546      |                   100 | mA                     | 112823.pdf~^~BC546~^~100~^~mA      |
 112823.pdf | BC547CZL1G |                   100 | mA                     | 112823.pdf~^~BC547CZL1G~^~100~^~mA |
 112823.pdf | BC547C     |                   100 | mA                     | 112823.pdf~^~BC547C~^~100~^~mA     | t
 112823.pdf | BC547BZL1G |                   100 | mA                     | 112823.pdf~^~BC547BZL1G~^~100~^~mA |
 112823.pdf | BC547CG    |                   100 | mA                     | 112823.pdf~^~BC547CG~^~100~^~mA    |
 112823.pdf | BC548      |                   100 | mA                     | 112823.pdf~^~BC548~^~100~^~mA      |
 112823.pdf | BC548BZL1G |                   100 | mA                     | 112823.pdf~^~BC548BZL1G~^~100~^~mA |
 112823.pdf | BC546BG    |                   100 | mA                     | 112823.pdf~^~BC546BG~^~100~^~mA    |
 112823.pdf | BRD8011    |                   100 | mA                     | 112823.pdf~^~BRD8011~^~100~^~mA    |
 112823.pdf | BC547BG    |                   100 | mA                     | 112823.pdf~^~BC547BG~^~100~^~mA    |
 112823.pdf | BC548CG    |                   100 | mA                     | 112823.pdf~^~BC548CG~^~100~^~mA    |
 112823.pdf | BC547BRL1G |                   100 | mA                     | 112823.pdf~^~BC547BRL1G~^~100~^~mA |
 112823.pdf | BC548BG    |                   100 | mA                     | 112823.pdf~^~BC548BG~^~100~^~mA    |
 112823.pdf | BC548CZL1G |                   100 | mA                     | 112823.pdf~^~BC548CZL1G~^~100~^~mA |
 112823.pdf | BC548BRL1G |                   100 | mA                     | 112823.pdf~^~BC548BRL1G~^~100~^~mA |
 112823.pdf | BC547AZL1G |                   100 | mA                     | 112823.pdf~^~BC547AZL1G~^~100~^~mA |
 112823.pdf | BC546B     |                   100 | mA                     | 112823.pdf~^~BC546B~^~100~^~mA     | t
 112823.pdf | BC547      |                   100 | mA                     | 112823.pdf~^~BC547~^~100~^~mA      |
 112823.pdf | BC546BRL1G |                   100 | mA                     | 112823.pdf~^~BC546BRL1G~^~100~^~mA |
 112823.pdf | BC548C     |                   100 | mA                     | 112823.pdf~^~BC548C~^~100~^~mA     | t
 112823.pdf | BC547ARLG  |                   100 | mA                     | 112823.pdf~^~BC547ARLG~^~100~^~mA  |
 112823.pdf | BC546BRL1  |                   100 | mA                     | 112823.pdf~^~BC546BRL1~^~100~^~mA  |
 112823.pdf | BC548B     |                   100 | mA                     | 112823.pdf~^~BC548B~^~100~^~mA     | t
(26 rows)
```

**Comment by lukehsiao** [comment]
Strange. My same view of that table is this:
```
lh@dd:~/repos/HardwareDeepDive$ deepdive sql 'select * from part_collector_current_candidates order by doc_id'
                       doc_id                        |    part_num     | collector_current_max | collector_current_unit |                                relation_id                                | is_true
-----------------------------------------------------+-----------------+-----------------------+------------------------+---------------------------------------------------------------------------+---------
 112823.pdf                                          | BC547BG         |                   100 | mA                     | 112823.pdf~^~BC547BG~^~100~^~mA                                           |
 112823.pdf                                          | BC547C          |                   100 | mA                     | 112823.pdf~^~BC547C~^~100~^~mA                                            | t
 112823.pdf                                          | BC548B          |                   100 | mA                     | 112823.pdf~^~BC548B~^~100~^~mA                                            | t
 112823.pdf                                          | BC547AZL1G      |                   100 | mA                     | 112823.pdf~^~BC547AZL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC546B          |                   100 | mA                     | 112823.pdf~^~BC546B~^~100~^~mA                                            | t
 112823.pdf                                          | BC547CZL1G      |                   100 | mA                     | 112823.pdf~^~BC547CZL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC548BZL1G      |                   100 | mA                     | 112823.pdf~^~BC548BZL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC547BZL1G      |                   100 | mA                     | 112823.pdf~^~BC547BZL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC546           |                   100 | mA                     | 112823.pdf~^~BC546~^~100~^~mA                                             |
 112823.pdf                                          | BC546           |                   100 | mA                     | 112823.pdf~^~BC546~^~100~^~mA                                             |
 112823.pdf                                          | BC546BG         |                   100 | mA                     | 112823.pdf~^~BC546BG~^~100~^~mA                                           |
 112823.pdf                                          | BC547A          |                   100 | mA                     | 112823.pdf~^~BC547A~^~100~^~mA                                            | t
 112823.pdf                                          | BC546B          |                   100 | mA                     | 112823.pdf~^~BC546B~^~100~^~mA                                            | t
 112823.pdf                                          | BC548BRL1G      |                   100 | mA                     | 112823.pdf~^~BC548BRL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC548CZL1G      |                   100 | mA                     | 112823.pdf~^~BC548CZL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC548           |                   100 | mA                     | 112823.pdf~^~BC548~^~100~^~mA                                             |
 112823.pdf                                          | BRD8011         |                   100 | mA                     | 112823.pdf~^~BRD8011~^~100~^~mA                                           |
 112823.pdf                                          | BC547BRL1G      |                   100 | mA                     | 112823.pdf~^~BC547BRL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC547           |                   100 | mA                     | 112823.pdf~^~BC547~^~100~^~mA                                             |
 112823.pdf                                          | BC548CG         |                   100 | mA                     | 112823.pdf~^~BC548CG~^~100~^~mA                                           |
 112823.pdf                                          | BC546BRL1       |                   100 | mA                     | 112823.pdf~^~BC546BRL1~^~100~^~mA                                         |
 112823.pdf                                          | BC546           |                   100 | mA                     | 112823.pdf~^~BC546~^~100~^~mA                                             |
 112823.pdf                                          | BC548BG         |                   100 | mA                     | 112823.pdf~^~BC548BG~^~100~^~mA                                           |
 112823.pdf                                          | BC546BRL1G      |                   100 | mA                     | 112823.pdf~^~BC546BRL1G~^~100~^~mA                                        |
 112823.pdf                                          | BC547ARLG       |                   100 | mA                     | 112823.pdf~^~BC547ARLG~^~100~^~mA                                         |
 112823.pdf                                          | BC548B          |                   100 | mA                     | 112823.pdf~^~BC548B~^~100~^~mA                                            | t
 112823.pdf                                          | BC547CG         |                   100 | mA                     | 112823.pdf~^~BC547CG~^~100~^~mA                                           |
 112823.pdf                                          | BC547A          |                   100 | mA                     | 112823.pdf~^~BC547A~^~100~^~mA                                            | t
 112823.pdf                                          | BC548C          |                   100 | mA                     | 112823.pdf~^~BC548C~^~100~^~mA                                            | t
 112823.pdf                                          | BC547ARL        |                   100 | mA                     | 112823.pdf~^~BC547ARL~^~100~^~mA                                          |
 112823.pdf                                          | BC546BZL1G      |                   100 | mA                     | 112823.pdf~^~BC546BZL1G~^~100~^~mA                                        |
```

Have you pulled from dev?

**Comment by xiao-cheng** [comment]
Yes I did pull after my last comment. I did run on the entire gold set, and the table is the same as before...I did notice some discrepancy between our runs before, it might be worthwhile to figure out what is causing this cross-platform issue.

**Comment by lukehsiao** [comment]
Ah, interesting. That would make it harder to debug. 

Do you think the version of psql and deepdive make a difference?

I'm running:
```
lh@dd:~/repos/HardwareDeepDive$ deepdive version
deepdive v0.7.1 (Linux x86_64)
Information on this build of deepdive follows.

  deepdive version: v0.7.1-0-g86cf6b7
  deepdive Git commit: 86cf6b79b12284b411d28ab04e7d8bb88047467f

  Build Date: 2015-09-28T11:41:20-07:00
  Build Host: raiders4.stanford.edu
  Build Operating System: Linux
  Build Machine: x86_64

  Running Operating System: Linux
  Running Machine: x86_64
```

and:
```
lh@dd:~/repos/HardwareDeepDive$ deepdive sql
psql (9.4.5)
Type "help" for help.
```
on a 64-bit Ubuntu 15.10 VM.

Can you show me what the calibration charts look like for you for both storage temp and for collector current when you run the gold set?

**Comment by xiao-cheng** [comment]
```
xiao@xiao-MBP ~/hdeepdive $ deepdive version
deepdive v0.7.1 (Darwin x86_64)
Information on this build of deepdive follows.

  deepdive version: v0.7.1-0-g86cf6b7
  deepdive Git commit: 86cf6b79b12284b411d28ab04e7d8bb88047467f

  Build Date: 2015-09-29T03:40:06+09:00
  Build Host: shin.local
  Build Operating System: Darwin
  Build Machine: x86_64

  Running Operating System: Darwin
  Running Machine: x86_64
```
and
```
xiao@xiao-MBP ~/hdeepdive $ psql
psql (9.4.4)
Type "help" for help.

xiao=#
```
Let me get the charts...

**Comment by xiao-cheng** [comment]
For some reason mine did not generate the image, just the tsv files, but histograms should be simple to read:

```
xiao@xiao-MBP ~/hdeepdive/run/LATEST/calibration $ cat part_collector_current.label.tsv
0.00	0.10	1	0	0
0.10	0.20	6	0	0
0.20	0.30	20	1	0
0.30	0.40	49	3	1
0.40	0.50	1446	22	29
0.50	0.60	577	17	19
0.60	0.70	5	0	0
0.70	0.80	0	0	0
0.80	0.90	0	0	0
0.90	1.00	0	0	0
xiao@xiao-MBP ~/hdeepdive/run/LATEST/calibration $ cat part_temperature.label.tsv
0.00	0.10	11	0	0
0.10	0.20	24	1	0
0.20	0.30	0	0	0
0.30	0.40	0	0	0
0.40	0.50	1	1	0
0.50	0.60	3	1	0
0.60	0.70	0	0	0
0.70	0.80	0	0	0
0.80	0.90	538	27	0
0.90	1.00	726	24	6
```

**Comment by lukehsiao** [comment]
Hmmmm. Those look different as well. Should I expect the calibration charts to look different between runs if no code was changed and `deepdive initdb` was run before each run?

**Comment by xiao-cheng** [comment]
I think the charts are different between each run due to the random nature of sampling algorithms. Giving guarantees on the result stability across runs is actually an active research topic in the group.

**Comment by lukehsiao** [comment]
I see. Well, I'm going to just leave this issue in the backlog then for now. I'm going to assume that me seeing duplicates isn't affecting #28. Thanks for all the quick help, Xiao!

**Comment by xiao-cheng** [comment]
I do think there is some error involved, because last time with the same code we extracted different number of pdfs. I think you can append debug value (such as current process id and the candidate hash etc) to the part number so you can see it in the output table without too much hacking

**Comment by lukehsiao** [comment]
Replacing the rel_id with the hash like this:
```python
        rel_id = hash('~^~'.join([x[PART_DOC_ID],
                             x[PART_NUM],
                             x[TEMP_MIN],
                             x[TEMP_MAX],
                             x[TEMP_UNIT]]))

        candidates.add('\t'.join(
                        [str(i) for i in[
                        x[PART_DOC_ID],
                        x[PART_NUM],
                        x[TEMP_MIN],
                        x[TEMP_MAX],
                        x[TEMP_UNIT],
                        rel_id,
                        is_true
                        ]]))
    for candidate in candidates:
        print candidate
```

Gives me this:
```
lh@dd:~/repos/HardwareDeepDive$ deepdive sql 'select * from part_temperature_candidates where is_true=true order by doc_id'
                                  doc_id                                   | part_num  | temp_min | temp_max | temp_unit |     relation_id      | is_true
---------------------------------------------------------------------------+-----------+----------+----------+-----------+----------------------+---------
 112823.pdf                                                                | BC548C    |      -55 |      150 | C         | -2132658710079284317 | t
 112823.pdf                                                                | BC548B    |      -55 |      150 | C         | 6043668724832557906  | t
 112823.pdf                                                                | BC546B    |      -55 |      150 | C         | -3120140003661737012 | t
 112823.pdf                                                                | BC546B    |      -55 |      150 | C         | -3120140003661737012 | t
[more stuff]
```
Where the duplicates have the same hash.

If I concatenate os.getpid() to the doc_id, I see this:
```
                                     doc_id                                     | part_num  | temp_min | temp_max | temp_unit |     relation_id      | is_true
--------------------------------------------------------------------------------+-----------+----------+----------+-----------+----------------------+---------
 15000+112823.pdf                                                                | BC548B    |      -55 |      150 | C         | 6043668724832557906  | t
 15000+112823.pdf                                                                | BC546B    |      -55 |      150 | C         | -3120140003661737012 | t
 15000+112823.pdf                                                                | BC547A    |      -55 |      150 | C         | -1415720250307987792 | t
 15003+112823.pdf                                                                | BC548C    |      -55 |      150 | C         | -2132658710079284317 | t
 15003+112823.pdf                                                                | BC548B    |      -55 |      150 | C         | 6043668724832557906  | t
 15003+112823.pdf                                                                | BC546B    |      -55 |      150 | C         | -3120140003661737012 | t
 15003+112823.pdf                                                                | BC547A    |      -55 |      150 | C         | -1415720250307987792 | t
 15003+112823.pdf                                                                | BC547C    |      -55 |      150 | C         | 8026941111078475850  | t
```

So it does look like multiple processes overlap.

**Comment by xiao-cheng** [comment]
I guess one simple solution is just to dedup the output tables or add UNIQUE as a constraint to the rel_id column, but we should definitely find out why two processes were invoked on the same file, maybe check the log for `deepdive initdb` and `deepdive run`

**Comment by lukehsiao** [comment]
So, it's not two processes running on the same file, it's two processes working on the same file, it's two processing working from the same tables.

```
part_collector_current_candidates +=
  ext_part_collector_current_candidates(doc_id,
                                        part_page_num,
                                        part_table_id,
                                        part_relation,
                                        part_relation_id,
                                        part_num,
                                        doc_id,
                                        cc_pg_num,
                                        cc_table_id,
                                        cc_relation,
                                        cc_relation_id,
                                        cc_max,
                                        cc_unit) :-
  part_num_mentions(doc_id, part_page_num, part_table_id, part_relation, part_relation_id, part_num),
  collector_current_mentions(doc_id, cc_pg_num, cc_table_id, cc_relation, cc_relation_id, cc_max, cc_unit).
```

This is just feeding info from collector_current_mentions and part_num_mentions to ext_part_collector_current_candidates. So with multiple processes, we should expect there to be some overlap.

**Comment by xiao-cheng** [comment]
I didn't quite understand. I thought the dependency between different tables is a tree where parent tables is computed once before children tables read into them.

**Comment by lukehsiao** [comment]
So, this bug was caused by us needing to maintain state inside our UDF for extracting candidates. For example, I needed to read in many rows of sys.stdin before creating outputs. Here is an excerpt of my chat with Jaeho on Gitter:

> Jaeho: That contract of udfs being stateless and being an independent function of each input row is quite fundamental. We exploit that property to parallelizing the data flow and you will surely get incorrect results if udf depends on the order of input etc. I recommend gathering all relevant input into a single row from the ddlog level. ARRAY_AGG()in Postgres may help to turn things into an array.

The [documentation on writing UDFs](http://deepdive.stanford.edu/writing-udf-python) for v0.8.0 also suggests that the UDF output 1 (or more) values for each line of input.

In contrast, we have UDFs which are parsing through *many* lines of input, building a data structure, then operating on it.

Because it is parallelized, we got duplicates.

I'm going to close this issue since I think we know what is causing it.

However, we still need to come up with a solution for this while we work on  #28. Otherwise our code will run too slowly and potentially incorrectly.
